Reactor plugin will cause app error

**Describe the bug**
A clear and concise description of what the bug is.
ound error
Caused by: java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread single-1
at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:83)
at reactor.core.publisher.Mono.block(Mono.java:1707)
at com.example.examplereactor.TestHook1.lambda$main$0(TestHook1.java:46)
at reactor.core.publisher.FluxStream.subscribe(FluxStream.java:48)
at reactor.core.publisher.InternalFluxOperator.subscribe(InternalFluxOperator.java:62)
at reactor.core.publisher.FluxSubscribeOn$SubscribeOnSubscriber.run(FluxSubscribeOn.java:194)
at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)

**Steps to reproduce**
If possible, provide a recipe for reproducing the error.
```java
package app.controller;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.CoreSubscriber;
import reactor.core.Scannable;
import reactor.core.publisher.Hooks;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Operators;
import reactor.core.scheduler.Schedulers;

import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;
import java.util.function.BiFunction;

public class TestHook {

    private static final Logger logger = LoggerFactory.getLogger(TestHook.class);

    static{
        try{
            Hooks.onEachOperator("testHook", Operators.lift(new Lifter<>()));
        }catch (Throwable ignored){
        }
    }


    public static class Lifter<T>
            implements BiFunction<Scannable, CoreSubscriber<? super T>, CoreSubscriber<? super T>> {

        /** Holds reference to strategy to prevent it from being collected. */

        @Override
        public CoreSubscriber<? super T> apply(Scannable publisher, CoreSubscriber<? super T> sub) {
            return sub;
        }
    }

    public static void main(String[] args) throws InterruptedException {
        Callable<String> callable = () -> {
            logger.info("Operation enter call1: ");
            Mono.just("test1").block();
            return "call1";
        };


        Mono.defer(() -> Mono.fromCallable(callable)
                             .publishOn(Schedulers.elastic())
                             .flatMap(source -> {
                                 logger.info("Again enter map :{} ", source);
                                 return Mono.just(source);
                             })).subscribeOn(Schedulers.single()).subscribe();


        TimeUnit.SECONDS.sleep(3);
    }
}

```
**What did you expect to see?**
A clear and concise description of what you expected to see.
no error found
**What did you see instead?**
A clear and concise description of what you saw instead.
https://github.com/reactor/reactor-core/issues/3450
https://github.com/reactor/reactor-core/issues/3452
**What version are you using?**
(e.g., `v0.9.0`, `393e4a2`, etc)
anyone will be reported it ,it's a bug of reactor
**Environment**
Compiler: (e.g., "AdoptOpenJDK 11.0.6")
OS: (e.g., "Ubuntu 20.04")
Runtime (if different from JDK above): (e.g., "Oracle JRE 8u251")
OS (if different from OS compiled on): (e.g., "Windows Server 2019")

**Additional context**
Add any other context about the problem here.
I have add a bug to reactor ,but they just suggest me to change my code ,but as javaagent developer ,we can not control app's code

--- a/instrumentation/reactor/reactor-3.1/library/src/main/java/io/opentelemetry/instrumentation/reactor/v3_1/ContextPropagationOperator.java
+++ b/instrumentation/reactor/reactor-3.1/library/src/main/java/io/opentelemetry/instrumentation/reactor/v3_1/ContextPropagationOperator.java
@@ -155,7 +155,8 @@ public final class ContextPropagationOperator {
 
   private static <T> Function<? super Publisher<T>, ? extends Publisher<T>> tracingLift(
       ReactorAsyncOperationEndStrategy asyncOperationEndStrategy) {
-    return Operators.lift(new Lifter<>(asyncOperationEndStrategy));
+    return Operators.lift(
+        ContextPropagationOperator::shouldInstrument, new Lifter<>(asyncOperationEndStrategy));
   }
 
   /** Forces Mono to run in traceContext scope. */
@@ -220,7 +221,12 @@ public final class ContextPropagationOperator {
     }
   }
 
-  public static class Lifter<T>
+  private static boolean shouldInstrument(Scannable publisher) {
+    // skip if Flux/Mono #just, #empty, #error
+    return !(publisher instanceof Fuseable.ScalarCallable);
+  }
+
+  private static class Lifter<T>
       implements BiFunction<Scannable, CoreSubscriber<? super T>, CoreSubscriber<? super T>> {
 
     /** Holds reference to strategy to prevent it from being collected. */
@@ -233,10 +239,6 @@ public final class ContextPropagationOperator {
 
     @Override
     public CoreSubscriber<? super T> apply(Scannable publisher, CoreSubscriber<? super T> sub) {
-      // if Flux/Mono #just, #empty, #error
-      if (publisher instanceof Fuseable.ScalarCallable) {
-        return sub;
-      }
       return new TracingSubscriber<>(sub, sub.currentContext());
     }
   }

--- a/instrumentation/reactor/reactor-3.1/library/src/test/java/io/opentelemetry/instrumentation/reactor/v3_1/HooksTest.java
+++ b/instrumentation/reactor/reactor-3.1/library/src/test/java/io/opentelemetry/instrumentation/reactor/v3_1/HooksTest.java
@@ -7,10 +7,14 @@ package io.opentelemetry.instrumentation.reactor.v3_1;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import org.junit.jupiter.api.Test;
 import reactor.core.CoreSubscriber;
+import reactor.core.Disposable;
 import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
 
 class HooksTest {
 
@@ -42,5 +46,29 @@ class HooksTest {
     public void subscribe(CoreSubscriber<? super Integer> actual) {
       subscriber.set(actual);
     }
+  }
+
+  @Test
+  void testInvalidBlockUsage() throws InterruptedException {
+    ContextPropagationOperator operator = ContextPropagationOperator.create();
+    operator.registerOnEachOperator();
+
+    Callable<String> callable =
+        () -> {
+          Mono.just("test1").block();
+          return "call1";
+        };
+
+    Disposable disposable =
+        Mono.defer(
+                () ->
+                    Mono.fromCallable(callable).publishOn(Schedulers.elastic()).flatMap(Mono::just))
+            .subscribeOn(Schedulers.single())
+            .subscribe();
+
+    TimeUnit.MILLISECONDS.sleep(100);
+
+    disposable.dispose();
+    operator.resetOnEachOperator();
   }
 }

